server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/HazelcastCluster.java

```diff
@@ -0,0 +1,246 @@
+    operationalProcessListenerUUID = operationalProcesses.addEntryListener(new OperationalProcessListener());
+  }
+
+  String getLocalUuid() {
+    return hzInstance.getLocalEndpoint().getUuid();
+  }
+
+  String getName() {
+    return hzInstance.getConfig().getGroupConfig().getName();
+  }
+
+  List<String> getMembers() {
+    return hzInstance.getCluster().getMembers().stream()
+      .filter(m -> !m.localMember())
+      .map(m -> m.getStringAttribute(HOSTNAME))
+      .collect(toList());
+  }
+
+  void addListener(@Nonnull AppStateListener listener) {
+    listeners.add(listener);
+  }
+
+  boolean isOperational(@Nonnull ProcessId processId) {
+    for (Map.Entry<ClusterProcess, Boolean> entry : operationalProcesses.entrySet()) {
+      if (entry.getKey().getProcessId().equals(processId) && entry.getValue()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void setOperational(@Nonnull ProcessId processId) {
+    operationalProcesses.put(new ClusterProcess(getLocalUuid(), processId), Boolean.TRUE);
+  }
+
+  boolean tryToLockWebLeader() {
+    IAtomicReference<String> leader = hzInstance.getAtomicReference(LEADER);
+    if (leader.get() == null) {
+      ILock lock = hzInstance.getLock(LEADER);
+      lock.lock();
+      try {
+        if (leader.get() == null) {
+          leader.set(getLocalUuid());
+          return true;
+        } else {
+          return false;
+        }
+      } finally {
+        lock.unlock();
+      }
+    } else {
+      return false;
+    }
+  }
+
+  public void registerSonarQubeVersion(String sonarqubeVersion) {
+    IAtomicReference<String> sqVersion = hzInstance.getAtomicReference(SONARQUBE_VERSION);
+    if (sqVersion.get() == null) {
+      ILock lock = hzInstance.getLock(SONARQUBE_VERSION);
+      lock.lock();
+      try {
+        if (sqVersion.get() == null) {
+          sqVersion.set(sonarqubeVersion);
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+
+    String clusterVersion = sqVersion.get();
+    if (!sqVersion.get().equals(sonarqubeVersion)) {
+      hzInstance.shutdown();
+      throw new IllegalStateException(
+        String.format("The local version %s is not the same as the cluster %s", sonarqubeVersion, clusterVersion)
+      );
+    }
+  }
+
+  @Override
+  public void close() {
+    if (hzInstance != null) {
+      // Removing listeners
+      operationalProcesses.removeEntryListener(operationalProcessListenerUUID);
+
+      // Removing the operationalProcess from the replicated map
+      operationalProcesses.keySet().forEach(
+        clusterNodeProcess -> {
+          if (clusterNodeProcess.getNodeUuid().equals(getLocalUuid())) {
+            operationalProcesses.remove(clusterNodeProcess);
+          }
+        });
+
+      // Shutdown Hazelcast properly
+      hzInstance.shutdown();
+    }
+  }
+
+  static Builder builder(ClusterProperties clusterProperties) {
+    return new Builder(clusterProperties);
+  }
+
+  static class Builder {
```

Comment:
This is not really a builder because this class instantiates a `HazelcastCluster` from a `ClusterProperties`. A single static method `create(ClusterProperties)` would be enough.